File: frontend/src/types/dicom.ts
/**
 * DICOM Type Definitions
 *
 * Core types for DICOM medical imaging data structures used
 * across the viewer, simulation, and segmentation modules.
 */

/** DICOM study metadata */
export interface DicomStudy {
  id: string;
  patientId: string;
  patientName: string;
  patientBirthDate: string;
  patientSex: string;
  studyInstanceUid: string;
  studyDate: string;
  studyTime: string;
  studyDescription: string;
  accessionNumber: string;
  referringPhysician: string;
  modalities: string[];
  seriesCount: number;
  instanceCount: number;
  createdAt: string;
  updatedAt: string;
}

/** DICOM series within a study */
export interface DicomSeries {
  id: string;
  studyId: string;
  seriesInstanceUid: string;
  seriesNumber: number;
  seriesDescription: string;
  modality: string;
  manufacturer: string;
  bodyPartExamined: string;
  laterality: string;
  protocolName: string;
  imageCount: number;
  seriesDate: string;
  rows: number;
  columns: number;
  sliceThickness: number;
  pixelSpacing: [number, number];
  windowCenter: number;
  windowWidth: number;
}

/** Viewport layout configuration */
export interface ViewportLayout {
  id: string;
  type: 'axial' | 'sagittal' | 'coronal' | '3d' | 'mpr';
  rows: number;
  columns: number;
}

/** MPR (Multi-Planar Reconstruction) view state */
export interface MPRViewState {
  axial: ViewportState;
  sagittal: ViewportState;
  coronal: ViewportState;
  crosshairPosition: [number, number, number];
  linked: boolean;
}

/** Individual viewport state */
export interface ViewportState {
  windowCenter: number;
  windowWidth: number;
  rotation: number;
  zoom: number;
  pan: [number, number];
  sliceIndex: number;
}

/** Window/Level preset */
export interface WindowLevelPreset {
  name: string;
  windowCenter: number;
  windowWidth: number;
  description: string;
}

/** Common CT window/level presets */
export const WINDOW_LEVEL_PRESETS: WindowLevelPreset[] = [
  { name: 'Lung', windowCenter: -600, windowWidth: 1500, description: 'Lung parenchyma' },
  { name: 'Mediastinum', windowCenter: 40, windowWidth: 400, description: 'Soft tissues' },
  { name: 'Bone', windowCenter: 300, windowWidth: 1500, description: 'Bone structures' },
  { name: 'Brain', windowCenter: 40, windowWidth: 80, description: 'Brain tissue' },
  { name: 'Abdomen', windowCenter: 50, windowWidth: 350, description: 'Abdominal organs' },
  { name: 'Liver', windowCenter: 60, windowWidth: 150, description: 'Liver parenchyma' },
];

File: backend/app/schemas/dicom.py
"""
DICOM Pydantic Schemas

Request/response schemas for DICOM study and series management APIs.
"""

from datetime import datetime
from pydantic import BaseModel, Field, field_validator
from typing import Optional, List


class DicomStudyResponse(BaseModel):
    """Response schema for DICOM study metadata."""
    id: str
    patient_id: str
    patient_name: str
    patient_birth_date: Optional[str] = None
    patient_sex: Optional[str] = None
    study_instance_uid: str
    study_date: Optional[str] = None
    study_time: Optional[str] = None
    study_description: Optional[str] = None
    accession_number: Optional[str] = None
    referring_physician: Optional[str] = None
    modalities: List[str] = []
    series_count: int = 0
    instance_count: int = 0
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True


class DicomSeriesResponse(BaseModel):
    """Response schema for DICOM series metadata."""
    id: str
    study_id: str
    series_instance_uid: str
    series_number: Optional[int] = None
    series_description: Optional[str] = None
    modality: Optional[str] = None
    manufacturer: Optional[str] = None
    body_part_examined: Optional[str] = None
    laterality: Optional[str] = None
    protocol_name: Optional[str] = None
    image_count: int = 0
    series_date: Optional[str] = None
    rows: Optional[int] = None
    columns: Optional[int] = None
    slice_thickness: Optional[float] = None
    pixel_spacing: Optional[List[float]] = None
    window_center: Optional[float] = None
    window_width: Optional[float] = None

    class Config:
        from_attributes = True

    @field_validator("pixel_spacing", mode="before")
    @classmethod
    def clean_pixel_spacing(cls, v):
        if v is None:
            return None
        if isinstance(v, list):
            cleaned = [x for x in v if x is not None]
            return cleaned if cleaned else None
        return v


class DicomInstanceResponse(BaseModel):
    """Response schema for DICOM instance metadata."""
    id: str
    series_id: str
    sop_instance_uid: str
    instance_number: Optional[int] = None
    slice_location: Optional[float] = None
    rows: Optional[int] = None
    columns: Optional[int] = None

    class Config:
        from_attributes = True


class DicomUploadResponse(BaseModel):
    """Response schema for DICOM upload operation."""
    study_id: str
    series_count: int
    instance_count: int
    message: str = "Upload successful"


class PaginatedResponse(BaseModel):
    """Generic paginated response wrapper."""
    items: List
    total: int
    page: int = 1
    page_size: int = 20
    total_pages: int = 0
